### DISCUSSION TITLE
Self-contained executable for Windows built on msys2

### DISCUSSION THREAD

**Discussion by efauk** [mpv-player/mpv#14375] [General]
Is it possible to produce a self-contained executable for Windows, when building on msys2? If so, what are the necessary link flags?
Thank you in advance.

**Comment by Andarwinux** [comment]
~~If you want a static exe, then you can't rely on prebuilt packages for msys2, you have to build everything from scratch yourself.~~

**Reply by kasper93** [reply]
Why? MSYS2 provides also static libs, it should be possible to produce fully static library.

**Reply by Andarwinux** [reply]
Oops, indeed. The last time I looked at msys2 there were still problems with ffmpeg's pkgconf and some dependencies, but that's been [fixed](https://github.com/msys2/MINGW-packages/pull/11689) for a while now.

**Reply by Andarwinux** [reply]
Doesn't seem a big problem, except that there is still hardcoded `-lstdc++` and `-lrt`.

**Comment by kasper93** [comment]
I never tried this with MSYS2, you can try `-Ddefault_library=static -Dprefer_static=true` in meson flags and check what happens.

**Reply by efauk** [reply]
Ok cool, that's why I thought initially, however I'm getting lots of "multiple definition" errors at linking stage.
I wonder if I need any other flag...

**Reply by kasper93** [reply]
> multiple definition

Which ones exactly?

**Reply by Andarwinux** [reply]
I'm guessing it's a rust stdlib [problem](https://github.com/Andarwinux/mpv/actions/runs/9553048011/job/26330959951) with libdovi and rav1e. mpv-winbuild-cmake is experiencing the same [issue](https://github.com/shinchiro/mpv-winbuild-cmake/issues/687), the cause of which is that libdovi.a and librav1e.a have different versions of stdlib embedded in them, our workaround is disable rav1e and replace it with svtav1.
So you can't rely on the ffmpeg provided by msys2 unless you build your own libplacebo with libdovi disabled.

**Reply by efauk** [reply]
I'm getting lots, one of them for instance is:

`
(...)/bin/ld.exe: (...)/msys64/ucrt64/lib/librav1e.a(rav1e.std-e80dc1a33fabde32.std.10161f498c56c4aa-cgu.06.rcgu.o.rcgu.o):(.text+0x3030): multiple definition of __rdl_alloc'; (...)/msys64/ucrt64/lib/librsvg-2.a(std-52a03da70d492f12.std.c7e05963787d85c4-cgu.04.rcgu.o):(.text+0x5bb0): first defined here`

**Reply by Andarwinux** [reply]
> I'm getting lots, one of them for instance is:
> 
> ` (...)/bin/ld.exe: (...)/msys64/ucrt64/lib/librav1e.a(rav1e.std-e80dc1a33fabde32.std.10161f498c56c4aa-cgu.06.rcgu.o.rcgu.o):(.text+0x3030): multiple definition of __rdl_alloc'; (...)/msys64/ucrt64/lib/librsvg-2.a(std-52a03da70d492f12.std.c7e05963787d85c4-cgu.04.rcgu.o):(.text+0x5bb0): first defined here`

You can try add `-Wl,--allow-multiple-definition` to LDFLAGS, which will force linker to ignore multiple definitions, but it's risky and you shouldn't use it in a production environment. Once rav1e add rust 1.79 support, this problem will go away on its own, so you can wait for a while.

**Reply by efauk** [reply]
I haven't managed to link correctly. I think some static libraries are missing, since I'm getting lots of undefined symbols now :(

I'm just copying a few, but I'm getting hundreds:

```
(...)/ld.exe: C:/msys64/ucrt64/lib/libSDL2.a(SDL_hidapi.c.obj):(.text+0x158f):
undefined reference to `__imp_SetupDiGetDeviceRegistryPropertyA'
(...)/ld.exe: C:/msys64/ucrt64/lib/libSDL2.a(SDL_windows_gaming_input.c.obj):(.
text+0xb54): undefined reference to `__imp_CM_Locate_DevNodeA'
(...)/ld.exe: C:/msys64/ucrt64/lib/libSDL2.a(SDL_windows_gaming_input.c.obj):(.
text+0xba4): undefined reference to `__imp_CM_Get_Parent'
(...)/ld.exe: C:/msys64/ucrt64/lib/libSDL2.a(SDL_windows_gaming_input.c.obj):(.
text+0xbc2): undefined reference to `__imp_CM_Get_Device_IDA'
```

**Reply by Andarwinux** [reply]
try add `-lole32 -luuid -lsetupapi` to LDFLAGS.

**Reply by efauk** [reply]
Yeah! That seems to work. Great, thanks a lot.
Although I still getting hundreds of errors. I guess I need to fix them by manually adding the dependencies to LDFLAGS one by one?

**Reply by Andarwinux** [reply]
Yes. But you can also figure out the problematic pkgconf to manually fix `Requires.private` or `Libs.private`, which is easy enough for sysroot closed build, but I'm not sure it's appropriate in a distro environment.
You might consider reporting such problems to msys2, but that might take some time.
